osu.Server.Spectator/Hubs/SpectatorHub.cs

```diff
@@ -30,13 +81,44 @@ public async Task BeginPlaySession(SpectatorState state)
 
         public async Task SendFrameData(FrameDataBundle data)
         {
-            await Clients.Group(GetGroupId(CurrentContextUserId)).UserSentFrames(CurrentContextUserId, data);
+            using (var usage = await GetOrCreateLocalUserState())
+            {
+                var score = usage.Item?.Score;
+                Debug.Assert(score != null);
+
+                score.ScoreInfo.Statistics = data.Header.Statistics;
+                score.ScoreInfo.MaxCombo = data.Header.MaxCombo;
+                score.ScoreInfo.Combo = data.Header.Combo;
+                score.Replay.Frames.AddRange(data.Frames);
+
+                await Clients.Group(GetGroupId(CurrentContextUserId)).UserSentFrames(CurrentContextUserId, data);
+            }
         }
 
         public async Task EndPlaySession(SpectatorState state)
         {
             using (var usage = await GetOrCreateLocalUserState())
+            {
+                var userScore = usage.Item?.Score;
+                Debug.Assert(userScore != null);
+
+                var now = DateTimeOffset.UtcNow;
+
+                userScore.ScoreInfo.Date = now;
+                var legacyEncoder = new LegacyScoreEncoder(userScore, null);
+
+                string path = Path.Combine(REPLAYS_PATH, now.Year.ToString(), now.Month.ToString(), now.Day.ToString());
+
+                Directory.CreateDirectory(path);
+
+                string filename = $"{now.ToUnixTimeSeconds()}-{CurrentContextUserId}-{userScore.ScoreInfo.BeatmapInfo.OnlineID}.osr";
```

Comment:
This filename construction is worrying me a bit there for several reasons:

* Primarily, no theoretically-unique component like a guid that would substantially decrease possibility of naming clashes/overwrites
* I'm also not sure about the second granularity of the timestamp (rather than something like msecs or even ticks), or the lack of a ruleset ID next to the beatmap ID (think converts)

Reply:
It’s all just a temporary thing, but adding ruleset sounds okay.

For a clash to happen the same user would have to play and submit twice in a single second. I’m not sure that should be possible.